java/algorithm/src/main/java/org/apache/arrow/algorithm/search/VectorRangeSearcher.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.search;
+
+import org.apache.arrow.algorithm.sort.VectorValueComparator;
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * Search for the range of a particular element in the target vector.
+ */
+public class VectorRangeSearcher {
+
+  /**
+   * Search for the first occurrence of an element.
+   * The search is based on the binary search algorithm. So the target vector must be sorted.
+   * @param targetVector the vector from which to perform the search.
+   * @param comparator the criterion for the comparison.
+   * @param keyVector the vector containing the element to search.
+   * @param keyIndex the index of the search key in the key vector.
+   * @param <V> the vector type.
+   * @return the index of the first matched element if any, and -1 otherwise.
+   */
+  public static <V extends ValueVector> int getLowerBound(
+          V targetVector, VectorValueComparator<V> comparator, V keyVector, int keyIndex) {
+    comparator.attachVectors(keyVector, targetVector);
+
+    int low = 0;
+    int high = targetVector.getValueCount() - 1;
+
+    while (low <= high) {
+      int mid = (low + high) / 2;
```

Comment:
overflow?  Didn't we already have binary search someplace else?

Reply:
Good suggestion for overflow.

This algorithm is similar to binary search. However, they solve different problems, and I think this one is more useful in practice. 

In particular, given a (sorted) vector with duplicated elements. This algorithm finds the bounds: the first/last element equal to the given element.

By different problems, I mean it is possible to solve the problem for binary search by this algorithm. But it is not efficient to solve the problem for this algorithm by binary search. 

To see this, note that binary search finds any match element, and after that, we may either go left-ward or right-ward to find the first/last match. However, the time complexity for the latter can be O(n), which is worse than the time complexity of this algorithm O(logn).